### PR TITLE
.Net: Include streaming tool call information in model diagnostics

### PR DESCRIPTION
**Pull request by TaoChenOSU**
### Motivation and Context

Tool call information is currently not included in the model diagnostics when using the streaming APIs.

### Description

1. Record OpenAI tool call information in model diagnostics for the streaming API.
2. If there is no tool call information, do not record an empty entry.

### Contribution Checklist

- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/ClientCore.cs [hunk 1]
```diff
@@ -717,10 +719,16 @@ static void AddResponseMessage(ChatCompletionsOptions chatOptions, ChatHistory c
                         streamedContents?.Add(openAIStreamingChatMessageContent);
                         yield return openAIStreamingChatMessageContent;
                     }
+
+                    // Translate all entries into ChatCompletionsFunctionToolCall instances.
+                    toolCalls = OpenAIFunctionToolCall.ConvertToolCallUpdatesToChatCompletionsFunctionToolCalls(
+                        ref toolCallIdsByIndex, ref functionNamesByIndex, ref functionArgumentBuildersByIndex);
+                    // Translate all entries into FunctionCallContent instances for diagnostics purposes.
+                    functionCallContents = activity is null ? null : toolCalls.Select(this.GetFunctionCallContent).ToArray();
```
Comment (stephentoub) [inline dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/ClientCore.cs:727]: Rather than doing all this translation, can we just get the original response from the Azure SDK object? Doesn't it make it available as a raw response content off of the relevant objects? It'd be both cheaper and higher fidelity.

Comment (stephentoub) [reply dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/ClientCore.cs:727]: Is checking `activity is null` the right check here, or do we need to check if the sensitive model diagnostics are enabled? It'd be good to avoid doing such work except in the case where it's actually going to be used.

File: dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/ClientCore.cs [hunk 2]
```diff
@@ -755,7 +759,17 @@ static void AddResponseMessage(ChatCompletionsOptions chatOptions, ChatHistory c
             // Add the original assistant message to the chatOptions; this is required for the service
             // to understand the tool call responses.
             chatOptions.Messages.Add(GetRequestMessage(streamedRole ?? default, content, streamedName, toolCalls));
-            chat.Add(new OpenAIChatMessageContent(streamedRole ?? default, content, this.DeploymentOrModelName, toolCalls, metadata) { AuthorName = streamedName });
+            // Add the result message to the caller's chat history
+            var newChatMessageContent = new OpenAIChatMessageContent(streamedRole ?? default, content, this.DeploymentOrModelName, toolCalls, metadata)
+            {
+                AuthorName = streamedName
+            };
+            // Add the tool call messages to the new chat message content for diagnostics purposes.
```
Comment (stephentoub) [inline dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/ClientCore.cs:767]: Why is this necessary when toolCalls is already being provided to the content ctor?

**Comment by stephentoub** [comment]
Do the other connectors that currently support function calling (Gemini, Mistral) have a similar problem?

File: dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/ClientCore.cs (hunk 1 above)
Comment (TaoChenOSU) [reply dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/ClientCore.cs:727]: The reason I didn't do that is because I would have to do the parsing inside the model diagnostic module, which is supposed to be agnostic to the model specifics.

The `FunctionCallContent` class carries enough information to represent the tool calls and is a common class. It's used in the non-streaming API, and AFAIK the Gemini connector is going to adapt to it soon.

File: dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/ClientCore.cs (hunk 2 above)
Comment (TaoChenOSU) [reply dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/ClientCore.cs:767]: It's because the message is of a derived type of the `ChatMessageContent` type, which doesn't not have the tool calls property.

**Comment by TaoChenOSU** [comment]
> Do the other connectors that currently support function calling (Gemini, Mistral) have a similar problem?

I intentionally did not record the tool calls information in the Gemini connector, since it's going to adapt to the `FunctionCallContent` and migrate to the new function call abstractions.

**Comment by TaoChenOSU** [comment]
> Do the other connectors that currently support function calling (Gemini, Mistral) have a similar problem?

Mistral has the same issue with its streaming API. I will address it in a separate PR.

File: dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/ClientCore.cs (hunk 1 above)
Comment (TaoChenOSU) [reply dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/ClientCore.cs:727]: > Is checking `activity is null` the right check here, or do we need to check if the sensitive model diagnostics are enabled? It'd be good to avoid doing such work except in the case where it's actually going to be used.

Nice catch! I will address this

File: dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/ClientCore.cs (hunk 2 above)
Comment (stephentoub) [reply dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/ClientCore.cs:767]: Isn't this going to interfere with non-diagnostic uses, though? This is the user's ChatHistory object, and this is now changing the ChatMessageContent to now have Items that it didn't previously have. That doesn't seem right. @SergeyMenshykh?

Comment (SergeyMenshykh) [reply dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/ClientCore.cs:767]: > Isn't this going to interfere with non-diagnostic uses, though? This is the user's ChatHistory object, and this is now changing the ChatMessageContent to now have Items that it didn't previously have. That doesn't seem right. @SergeyMenshykh?

@TaoChenOSU, do you use the FunctionCallContent items from the chat message content items for model diagnostics? Or are you adding them because similar functionality exists for non-streaming APIs?  
   
Either way, this change will be required to support function call content classes for streaming. I would have added it anyway, within the scope of the task - https://github.com/microsoft/semantic-kernel/issues/5818, to communicate the fact of the function call by LLM to callers in a connector-agnostic manner.  
   
I don't expect callers to be impacted by the change. On the contrary, they will be able to get function call info from the chat message item's Items collection in a connector-agnostic way rather than casting the chat message to the OpenAI one and accessing the info via the ToolCalls property.

Comment (TaoChenOSU) [reply dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/ClientCore.cs:767]: It’s required for model diagnostics.

File: dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/ClientCore.cs [hunk 3]
```diff
@@ -717,10 +719,16 @@ static void AddResponseMessage(ChatCompletionsOptions chatOptions, ChatHistory c
                         streamedContents?.Add(openAIStreamingChatMessageContent);
                         yield return openAIStreamingChatMessageContent;
                     }
+
+                    // Translate all entries into ChatCompletionsFunctionToolCall instances.
+                    toolCalls = OpenAIFunctionToolCall.ConvertToolCallUpdatesToChatCompletionsFunctionToolCalls(
+                        ref toolCallIdsByIndex, ref functionNamesByIndex, ref functionArgumentBuildersByIndex);
+                    // Translate all entries into FunctionCallContent instances for diagnostics purposes.
+                    functionCallContents = ModelDiagnostics.IsSensitiveEventsEnabled() ? toolCalls.Select(this.GetFunctionCallContent).ToArray() : null;
```
Comment (markwallace-microsoft) [inline dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/ClientCore.cs:727]: Do we need to check if we have an `activity`

Comment (TaoChenOSU) [reply dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/ClientCore.cs:727]: That's implicitly checked. The `ModelDiagnostics` module makes sure that the activity is null when model diagnostics is disabled. And `IsSensitiveEventEnabled` checks if model diagnostics is enabled and the collection of sensitive event is enabled.

Comment (SergeyMenshykh) [inline dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/ClientCore.cs:727]: The functionCallContents variable should always be initialized by the value returned by GetFunctionCallContent, regardless of whether logging of sensitive events is enabled or not. Otherwise, callers will have to enable sensitive events to have the function call content items in the chat message.  
   
The "if" logic can be moved a few lines below, where the list is passed to the EndStreaming method. Depending on the IsSensitiveEventsEnabled, either the functionCallContents value or null will be passed as an argument.  
   
Since the function call content classes for streaming are not officially supported yet, this behavior may be acceptable for now and can be changed later when adding call content classes for streaming.

Comment (TaoChenOSU) [reply dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/ClientCore.cs:727]: Agreed. I think we can remove this when we have more consistency across the connectors. I may modify this behavior when I work on the Mistral connector.

**Review by markwallace-microsoft — APPROVED** [review]
